### ISSUE TITLE
Lazarus

### ISSUE THREAD

**Issue by Erion57** [shtrih-m/fr_drv_ng#281]
Уважаемые разработчики, убедительная просьба добавить минимальный пример работы в Lazarus на linux в fr_drv_ng_examples.
Или хотя бы подскажите как подключить драйвер и минимально выполнить connect ну и beep. Спасибо.

**Comment by swex** [comment]
Здравствуйте,
библиотека написана на C++, готового способа трансляции C++ в .pas нет.

Вы можете посмотреть [здесь](https://forum.lazarus.freepascal.org/index.php?topic=49101.0) описаны различные способы взаимодействия. 

Так же в состав C++ библиотеки входит C интерфейс ([пример](https://github.com/shtrih-m/fr_drv_ng_examples/blob/master/classic_interface/c_main.cpp)), вот его проще будет адаптировать под FreePascal. Советуют [h2pas](https://www.freepascal.org/tools/h2pas.html)

Еще, можете посмотреть в сторону[json-rpc](https://github.com/shtrih-m/fr_drv_ng_examples/tree/master/classic_jsonrpc), в таком случае вы сможете работать без линковки с библиотекой, но и обмен будет по http/websocket json сообщениями, это может быть неудобно.

**Comment by Erion57** [comment]
А возможен вариант использовать external library и подключить so файл драйвера? Если возможно, то не совсем понимаю как при этом с ним взаимодействовать.

**Comment by swex** [comment]
Возможно. Вот [здесь](https://wiki.freepascal.org/Creating_bindings_for_C_libraries) описано "как создавать биндинги".
Вот заголовочный, который поможет вам сгенерировать обертку
[c_classic_interface.zip](https://github.com/shtrih-m/fr_drv_ng/files/14795637/c_classic_interface.zip)

Вот [документация](https://wiki.freepascal.org/Creating_bindings_for_C_libraries) как генерировать обертки от самого Lazarus

Вот [пример](https://github.com/shtrih-m/fr_drv_ng_examples/blob/master/classic_interface/c_main.cpp), правда на C++ как библиотека подгружается динамически через dlopen и C-linkage функции(методы, геттеры/сеттеры) выцепляются из so в рантайме по имени

**Comment by Erion57** [comment]
Все получилось. Только вопрос, например get_ECRModeDescription возвращает "Day closed" и т.п., это правильно или можно как то переключить на русский? А в Get_SerialNumber лидирующий 0 пропал, в чем может быть дело?

**Comment by swex** [comment]
По умолчанию берется из системной локали.
```bash
LC_ALL=en_US.utf-8 console_test_fr_drv_ng status #сообщения на английском
LC_ALL=ru_RU.utf-8 console_test_fr_drv_ng status #сообщения на русском
```
так же можно принудительно выставить  через собственную переменную окружения `FR_DRV_LANG` можно принудительно переключить локаль. Вот так:
```bash
export FR_DRV_LANG=ru
console_test_fr_drv_ng status
```

**Comment by swex** [comment]
>  А в Get_SerialNumber лидирующий 0 пропал, в чем может быть дело?

Без лога не могу сказать точно в чем там дело, но по этому поводу скажу только что получать ЗН через запрос статуса не верно, правильный путь сейчас - это вычитывать ЗН из таблиц.

**Comment by Erion57** [comment]
А метод ShowProperties как то по другому стал работать? Вызываю и ничего не происходит, в ответ получаю " -12 "

**Comment by swex** [comment]
> А метод ShowProperties как то по другому стал работать? Вызываю и ничего не происходит, в ответ получаю " -12 "

В этом драйвере нет методов, завязанных на графический интерфейс. Всё можно настроить установкой свойств.

**Comment by Erion57** [comment]
Понял, спасибо, просто метод то есть)

**Comment by swex** [comment]
> Понял, спасибо, просто метод то есть)

Да, много методов, которые будут возвращать, -12 - Не реализовано. Сделано для совместимости
